### ISSUE TITLE
RENDERERGL2: Strange shadows on skyboxes when looking through glass shaders

### ISSUE THREAD

**Issue by kai-li-wop** [PadWorld-Entertainment/worldofpadman#266]
One very strange thing is that the skybox gets shadows, but only when looking through a glass shader. Since WoP works a lot with glass shaders, I think this affects the user experience a lot when using the OpenGL2 renderer. Screenshots are take from PadShip but it also occurs in many other maps like CybJourney. Maybe @zturtleman could have a look because this should also appear with ioquake3. See example shader how it is used in PadShip to create the glass effect.

```
textures/pad_pirate/pglass04
{
	qer_editorimage textures/pad_pirate/pglass04
	surfaceparm trans
	cull none
	surfaceparm nolightmap
	qer_trans 	0.5

	{
		map textures/pad_pirate/pglass04
		blendFunc GL_ONE GL_ONE
		rgbGen identity
	}
	{
		map $lightmap
		rgbGen identity
		blendFunc filter
	}
}
```

![grafik](https://github.com/PadWorld-Entertainment/worldofpadman/assets/44040989/d41069cc-0629-4bf4-b4f8-af08cd675de6)

![grafik](https://github.com/PadWorld-Entertainment/worldofpadman/assets/44040989/c7ff8e8b-8b12-42cd-a084-dca788bbbe65)

**Comment by zturtleman** [comment]
I looked into this a month ago and I don't know what's going on with the shadows. Rotating the player left/right rotates the shadows in front of you like a clock. Wrong shadow projection/axis maybe? I don't really know how the shadow mapping works.

However it seems like all the glass shaders have a lightmap stage and `surfaceParm nolightmap`. I think nolightmap forces lightmapIndex to LIGHTMAP_BY_VERTEX which uses white image for lightmap? The opengl2 renderer could probably disable sun shadows on shaders with `surfaceParm nolightmap` (or possibly negative lightmapIndex) to fix the issue. I haven't got round to specifically testing and committing a solution to ioquake3 though.

**Comment by zturtleman** [comment]
I fixed the issue in ioquake3 by changing the shadow mapping instead of my previous suggested workaround to disable it on these shaders that happen to request no lightmap. https://github.com/ioquake/ioq3/pull/680

**Comment by mgerhardy** [comment]
Thanks a lot - cherry-picked into world of padman with 8eaf5e560bb32724af4d4bc5033b82513a0341fb

**Comment by kai-li-wop** [comment]
@zturtleman: It seems not to be fixed. The shadows appear different but are still there. See picture taken in wop_journey. Same in wop_padship. Not sure, did we miss something to pick from ioquake3 @mgerhardy?

![Image](https://github.com/user-attachments/assets/9e2e82b6-fe06-42f2-85b3-0b797b2563a3)

**Comment by mgerhardy** [comment]
Nope - everything is picked - just verified again. ~~Where is it in `wop_padship`? The one from the previous screenshot was fixed for me.~~

UPDATE: still visible for me, too - just not in the same way and the same window I've checked before.

**Comment by zturtleman** [comment]
Using using the latest wop code, I don't see the shadow in wop_journeyCTL. I think I checked all the windowed but what is the location from `viewpos` console command?

To debug this, change this line to `if (1) {` to display the depth and screen-space shadow image on the HUD. If wop_padship has a second ship in the sky in the shadow image like the [first image here](https://github.com/ioquake/ioq3/pull/680), it's the same problem. https://github.com/PadWorld-Entertainment/worldofpadman/blob/e1724311838f33a4a764b1f91b6b2277b98a2962/code/renderergl2/tr_backend.c#L1536-L1542

I guessing the epsilon need to be increased. Maybe try `if ( depth >= 1.0 - DEPTH_MAX_ERROR * 2 )`.
 https://github.com/PadWorld-Entertainment/worldofpadman/blob/8eaf5e560bb32724af4d4bc5033b82513a0341fb/code/renderergl2/glsl/shadowmask_fp.glsl#L106

I just used the existing DEPTH_MAX_ERROR constant in the file as an epsilon as it worked for me. I'm using AMD graphic card on Linux. Nvidia or different driver maybe have different precision and need a larger epsilon. (or I'm looking at the wrong window.)

**Comment by kai-li-wop** [comment]
I added some screenshots with the viewpos coordinates. I hope this is suitable. I am running WoP on Windows 10 with some old AMD Radeon.

GPU - AMD Radeon R9 200 Series
VRAM - 8192 MB - GDDR5 1375 MHz
Memory Clock - 1375 MHz
Core Clock - 1030 MHz
Total Memory Bandwidth - 352 GByte/s
Memory Bit Rate - 5.50 Gbps
OpenGL® API Version - 4.6
OpenCL™ API Version - 2.0

![grafik](https://github.com/user-attachments/assets/5f644787-a69a-4e82-8bc1-cd6161e702a6)

![grafik](https://github.com/user-attachments/assets/5328494a-cc05-4f38-adb4-73351a49cf01)

![grafik](https://github.com/user-attachments/assets/4604b1aa-fd7e-4bc2-87c1-ffbae21a42f1)

![grafik](https://github.com/user-attachments/assets/59912ce7-9423-4fa8-a1db-46a7994bb828)

Looks like there is a second ship. ;)

![grafik](https://github.com/user-attachments/assets/35c926b3-693c-4101-9407-457e854a3671)

**Comment by kai-li-wop** [comment]
Looks like `if ( depth >= 1.0 - DEPTH_MAX_ERROR * 2 )` doesn't change anything. I am not sure if I do something wrong while compiling.

![grafik](https://github.com/user-attachments/assets/6a342c15-fdd8-433d-a32f-e942f0261dc3)

**Comment by zturtleman** [comment]
Okay, I'm definitely not seeing the incorrect shadows in wop_journeyCTL and seems to be the same issue with shadow image in the screenshots.

Times 2 was a guess, I don't have anything to properly base it on. Try `if ( depth >= 1.0 - 0.0000001 )` and if it doesn't work try removing a 0 in 0.**000000**1 until it works. (Want the most 0s that works or it might remove shadows on far geometry.)

**Comment by kai-li-wop** [comment]
`if ( depth >= 1.0 - 0.0001 )` works on wop_padship, when looking through the window (1st screenshot). But when I float (noclip) through the window far enough away, it's coming back but not visible for the normal player (2nd screenshot). But this is still not working on wop_journeyCTL. Needs further testing.

![grafik](https://github.com/user-attachments/assets/c6df5d5c-c74e-4045-8938-679d9cd2f3e5)
![grafik](https://github.com/user-attachments/assets/a4093cf4-d7a4-4be9-b597-520b1d562e9b)

**Comment by kai-li-wop** [comment]
`if ( depth >= 1.0 - 0.001 )` (one zero less) also removes the issue on wop_journeyCTL but also removes the possibility to see brushwork and skybox while looking through two glass window shaders. Compare left and right part of the window on the screenshot. But the issue is gone on wop_padship for sure now. Also no shadow when floating far outside the window, so **0.001** it is then.

![grafik](https://github.com/user-attachments/assets/93256997-e096-4f82-a7b6-adbee8255ed4)

**Comment by kai-li-wop** [comment]
I need to correct myself. The issues are not connected. It is also not possible to see the skybox/brushwork while looking through 2 glass window shaders with current code `if ( depth >= 1.0 - DEPTH_MAX_ERROR )`. I guess due to the blend function of the shader the brightness simply adds up to plain white.

![grafik](https://github.com/user-attachments/assets/43f7c4cc-8b1d-489d-a900-2244e279c79c)

**Comment by zturtleman** [comment]
Okay, it seems reasonable to me. Before I merge https://github.com/ioquake/ioq3/pull/684, are you done testing?

I changed it from `if ( depth >= 1.0 - 0.001 )` to `if ( depth >= 0.999 )`. It should be mathematically _similar_, floating point can round different. I don't imagine it would cause a problem but if could briefly check again it's okay I would appreciate it. (I should of thought of this before, sorry.)

**Comment by kai-li-wop** [comment]
`if ( depth >= 0.999 )` just works fine on my side, no ghost shadows visible. Maybe @mgerhardy and @3aTmE could approve this fix on their machines?

And thank you for taking care of this problem. This was the most visible drawback of the openGL2 renderer in WoP.

**Comment by kai-li-wop** [comment]
Should be fixed with 63ab631c948f50a154bbf7ae755c119cedb0fe14

**Comment by 3aTmE** [comment]
I just tried it with the recent Codebase from Oct. 24. and was not able to recreate these strange shadows. 
This is seems definitely fixed. 
![Image](https://github.com/user-attachments/assets/0c7ab3de-c1ad-4e8c-9338-5f702059d166)

Though, there was something else i have noticed.
In the Captains Quarter are the shadows cast by the Sunlight only seen from the _outside_ not from the inside tho. 

**Inside:** 
![Image](https://github.com/user-attachments/assets/98eb2984-0bb4-495c-bf2d-715e0eba0e68)

**View from the Outside to the Inside:** 
![Image](https://github.com/user-attachments/assets/5df5cef3-9d9b-4154-858c-f05d3c970851)

I don't know why the Shadows are behaving like this.

**TLDR:** The Issue itself is fixed, shadows cast by the Sun are still quite iffy. :)

**System Info:**
Fedora 40
Wayland
AMD RTX 580 / AMD Ryzen 5 3600 CPU
32GB Ram

**Comment by kai-li-wop** [comment]
Shadows cast ok in the captain's room on my side. Visible through the window and while being inside of the room. But it looks on your screenshot like there is a little sun at the window bench but not at the table. I will keep this issue stated as fixed. We will see if more issues like this will occur during the beta tests.

![grafik](https://github.com/user-attachments/assets/cdcb1d7b-ae8a-4514-baf4-0878107bedc1)

**Comment by 3aTmE** [comment]
> Shadows cast ok in the captain's room on my side. Visible through the window and while being inside of the room. But it looks on your screenshot like there is a little sun at the window bench but not at the table. I will keep this issue stated as fixed. We will see if more issues like this will occur during the beta tests.
> 

Just checked it on my Laptop running Manjaro Linux with an Nvidia GPU just to be sure that this issue is nothing distro or driver related. Well, it pretty much is. 

On my Laptop the shadows also show up fine like they're supposed to. :)
